File: backend-springboot/src/main/java/com/monitor/controller/ApiMonitorController.java
package com.monitor.controller;

import com.monitor.service.ApiMonitorService;
import com.monitor.service.ElasticsearchService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

/**
 * 接口监控控制器
 */
@RestController
@RequestMapping("/api")
@RequiredArgsConstructor
public class ApiMonitorController {
    
    private final ApiMonitorService apiMonitorService;
    private final ElasticsearchService elasticsearchService;
    
    @Value("${default-project-id:001}")
    private String defaultProjectId;
    
    @PostMapping("/report")
    public ResponseEntity<Map<String, Object>> report(@RequestBody Map<String, Object> body) {
        return handleReport(body);
    }
    
    @GetMapping("/report")
    public ResponseEntity<Map<String, Object>> reportGet(@RequestParam String data) {
        try {
            String decoded = java.net.URLDecoder.decode(data, java.nio.charset.StandardCharsets.UTF_8);
            com.fasterxml.jackson.databind.ObjectMapper mapper = new com.fasterxml.jackson.databind.ObjectMapper();
            Map<String, Object> body = mapper.readValue(decoded, Map.class);
            return handleReport(body);
        } catch (Exception e) {
            System.err.println("API report GET failed: " + e.getMessage());
            return ResponseEntity.ok(Map.of("success", false, "message", "数据解析失败"));
        }
    }
    
    private ResponseEntity<Map<String, Object>> handleReport(Map<String, Object> body) {
        try {
            // 调试日志：检查 userId 是否正确传递
            if (body.get("userId") != null) {
                System.out.println("API monitor report - userId: " + body.get("userId") + ", url: " + body.get("url"));
            }
            
            apiMonitorService.report(
                defaultProjectId,
                (String) body.getOrDefault("url", ""),
                (String) body.getOrDefault("method", "GET"),
                body.get("status") != null ? ((Number) body.get("status")).intValue() : null,
                body.get("responseTime") != null ? ((Number) body.get("responseTime")).longValue() : null,
                (String) body.get("userId"),
                (String) body.get("sessionId"),
                body.get("requestData"),
                body.get("responseData")
            );
            
            // 异步写入 Elasticsearch（不阻塞主流程）
            Map<String, Object> logData = new HashMap<>();
            logData.put("projectId", defaultProjectId);
            logData.put("type", "api");
            logData.put("userId", body.get("userId"));
            logData.put("sessionId", body.get("sessionId"));
            logData.put("url", body.getOrDefault("url", ""));
            logData.put("timestamp", body.getOrDefault("timestamp", new java.util.Date().toInstant().toString()));
            logData.put("method", body.getOrDefault("method", "GET"));
            logData.put("status", body.get("status"));
            logData.put("responseTime", body.get("responseTime"));
            logData.put("message", body.getOrDefault("method", "GET") + " " + body.getOrDefault("url", "") + " - " + body.getOrDefault("status", 0) + " (" + body.getOrDefault("responseTime", 0) + "ms)");
            logData.put("requestData", body.get("requestData"));
            logData.put("responseData", body.get("responseData"));
            
            try {
                elasticsearchService.writeLog(logData);
            } catch (Exception e) {
                System.out.println("⚠️ Elasticsearch write failed (non-blocking): " + e.getMessage());
            }
            
            return ResponseEntity.ok(Map.of("success", true));
        } catch (Exception e) {
            System.err.println("API monitor report failed: " + e.getMessage());
            return ResponseEntity.ok(Map.of("success", false, "message", "上报失败"));
        }
    }
    
    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> stats() {
        Map<String, Object> stats = apiMonitorService.getStats(defaultProjectId);
        return ResponseEntity.ok(Map.of("success", true, "data", stats));
    }
    
    @GetMapping("/errors")
    public ResponseEntity<Map<String, Object>> getErrorDetails(
            @RequestParam String url,
            @RequestParam(required = false, defaultValue = "GET") String method) {
        // TODO: 实现接口错误详情查询
        return ResponseEntity.ok(Map.of(
            "success", true,
            "data", java.util.List.of()
        ));
    }
}

File: backend-midway/src/controller/api-monitor.controller.ts
import { Controller, Get, Post, Inject, Body, Query, Config } from '@midwayjs/core';
import { ApiTags, ApiOperation, ApiQuery, ApiBody } from '@midwayjs/swagger';
import { ApiMonitorService } from '../service/api-monitor.service';
import { ElasticsearchService } from '../service/elasticsearch.service';

@Controller('/api')
@ApiTags(['api'])
export class ApiMonitorController {
  @Inject()
  apiMonitorService: ApiMonitorService;

  @Inject()
  elasticsearchService: ElasticsearchService;

  @Config('defaultProjectId')
  defaultProjectId: string;

  @Post('/report')
  @ApiOperation({ summary: '上报接口监控数据' })
  @ApiBody({ description: '接口调用数据' })
  async report(@Body() body: any) {
    return this.handleReport(body);
  }

  @Get('/report')
  @ApiOperation({ summary: '上报接口监控数据（GET方式，用于img上报）' })
  @ApiQuery({ name: 'data', required: true, description: 'JSON编码的接口数据' })
  async reportGet(@Query('data') dataStr: string) {
    try {
      const body = JSON.parse(decodeURIComponent(dataStr));
      return this.handleReport(body);
    } catch (error) {
      console.error('API monitor report GET failed:', error);
      return { success: false, message: '数据解析失败' };
    }
  }

  private async handleReport(body: any) {
    try {
      // 调试日志：记录接收到的数据
      if (process.env.NODE_ENV === 'local' || process.env.NODE_ENV === 'development') {
        console.log('📥 API monitor report received:', {
          projectId: body.projectId,
          defaultProjectId: this.defaultProjectId,
          url: body.url,
          method: body.method,
          status: body.status,
          userId: body.userId,
          sessionId: body.sessionId
        });
      }
      
      await this.apiMonitorService.report({
        projectId: this.defaultProjectId,
        url: body.url || '',
        method: body.method || 'GET',
        status: body.status || 0,
        responseTime: body.responseTime || 0,
        userId: body.userId,
        sessionId: body.sessionId,
        requestData: body.requestData,
        responseData: body.responseData,
      });
      
      // 调试日志：记录数据上报成功
      if (process.env.NODE_ENV === 'local' || process.env.NODE_ENV === 'development') {
        console.log('✅ API monitor report saved successfully');
      }

      // 异步写入 Elasticsearch（不阻塞主流程）
      this.elasticsearchService.writeLog({
        projectId: this.defaultProjectId,
        type: 'api',
        userId: body.userId,
        sessionId: body.sessionId,
        url: body.url || '',
        timestamp: body.timestamp || new Date().toISOString(),
        method: body.method || 'GET',
        status: body.status || 0,
        responseTime: body.responseTime || 0,
        message: `${body.method || 'GET'} ${body.url || ''} - ${body.status || 0} (${body.responseTime || 0}ms)`,
        requestData: body.requestData,
        responseData: body.responseData,
      }).catch((err) => {
        // Elasticsearch 写入失败不影响主流程，只记录错误
        console.warn('⚠️ Elasticsearch write failed (non-blocking):', err.message);
      });

      return { success: true };
    } catch (error) {
      console.error('API monitor report failed:', error);
      return { success: false, message: '上报失败' };
    }
  }

  @Get('/stats')
  @ApiOperation({ summary: '获取接口监控统计' })
  async stats() {
    const stats = await this.apiMonitorService.getStats(this.defaultProjectId);
    return { success: true, data: stats };
  }

  @Get('/errors')
  @ApiOperation({ summary: '获取接口错误详情' })
  @ApiQuery({ name: 'url', required: true, description: '接口URL' })
  @ApiQuery({ name: 'method', required: false, description: '请求方法，默认GET' })
  async getErrorDetails(@Query('url') url: string, @Query('method') method: string = 'GET') {
    const errors = await this.apiMonitorService.getErrorDetails(
      this.defaultProjectId,
      url,
      method
    );
    return { success: true, data: errors };
  }
}

